### PR TITLE
Add SAC-driven fully overlapped activation CPU offload (primarily aimed at GH200/GB200/GB300 superchips) 

### PR DESCRIPTION
**Pull request by alint77**
# SAC-Driven Activation CPU Offload

## Summary

This PR adds a `torch.compile` path for activation CPU offload driven by
Selective Activation Checkpointing policies:

```python
CheckpointPolicy.MUST_CPU_OFFLOAD
CheckpointPolicy.PREFER_CPU_OFFLOAD
```

The intent is that users describe what should be offloaded through SAC policy,
and the compiler handles the rest. They should not need to call AO ops directly,
install a custom AOTAutograd pass, pick a prefetch window, or tune CPU/GPU
bandwidth constants.

Under compile, supported CPU-offload policies lower to async D2H/H2D copies with
explicit `ao.offload`, `ao.reload`, and `ao.wait_tensor` nodes. Eager mode gets a
correctness-first fallback that saves selected tensors on CPU and reloads them
when needed, but eager is not trying to provide async overlap.

Unsupported `MUST_CPU_OFFLOAD` cases fail clearly. Unsupported
`PREFER_CPU_OFFLOAD` cases fall back safely.

## User API

The public API is SAC policy:

```python
from functools import partial

import torch
from torch.utils.checkpoint import (
    CheckpointPolicy,
    checkpoint,
    create_selective_checkpoint_contexts,
)


def policy_fn(ctx, op, *args, **kwargs):
    if op == torch.ops.aten.mm.default:
        return CheckpointPolicy.MUST_CPU_OFFLOAD
    return CheckpointPolicy.PREFER_RECOMPUTE


context_fn = partial(create_selective_checkpoint_contexts, policy_fn)


def model(x, w):
    return checkpoint(
        lambda x, w: x @ w,
        x,
        w,
        use_reentrant=False,
        context_fn=context_fn,
    )


compiled_model = torch.compile(model)
```

### Checkpointing and partitioning

`torch.utils.checkpoint` now treats `MUST_CPU_OFFLOAD` and
`PREFER_CPU_OFFLOAD` as save policies.

For compiled graphs, AOTAutograd sees those policies and automatically invokes
activation offload. The partitioner marks offloaded values as saved across the
forward/backward boundary through their CPU copy, rather than also saving the
original GPU tensor or trying to recompute it.

The partitioner also walks forward descendants of offloaded values. If a
recompute-tagged node would need an activation that has been moved to CPU, that
node is forced to save instead. This avoids producing graphs that look valid on
paper but require a no-longer-resident GPU tensor in backward.

### AO runtime

The runtime keeps the explicit AO IR, but tightens the stream and lifetime model:

- D2H offload runs on a transfer stream after waiting for compute.
- The source GPU tensor records the transfer stream so the caching allocator
  cannot reuse storage while the copy is reading it.
- H2D reload allocates the destination on the compute stream, records it on the
  transfer stream, and waits before writing into it.
- `ao.wait_tensor` makes compute wait at the first real use.
- waits are keyed by device and data pointer, with loud errors for missing or
  duplicate waits.
- forward waits carry `last_use_of_storage` so Inductor's planned buffer reuse
  cannot overwrite storage still being read by async D2H.

This moves away from relying on `resize_(0)` as the lifetime mechanism.

### Scheduling

The old bandwidth-estimated reload prefetch path is removed. There is no
user-facing prefetch window and no `activation_offload_cpu_gpu_bw` tuning knob.

The new schedule is structural:

- forward waits are pipelined with a small bounded distance instead of all being
  sunk to graph output.
- backward reloads use a fixed one-step pipeline: reload the next activation
  after the current offloaded activation reaches its wait/first-use point.
- rewritten graphs are validated for topological order and one wait per
  transfer.

This came from trace debugging: loose placement can accidentally cluster all
H2Ds together, which destroys the intended overlap.

### Compiler ordering

The PR also includes the Inductor ordering fixes needed for the AO schedule to
survive lowering and scheduling:

- flatten nested `control_deps` dependency containers.
- treat `control_deps` dependencies as ordering-only in reinplace.
- expand dict-returning Triton wrapper deps when needed.
- model pass-through `control_deps` outputs with `MutationOutput` instead of
  fake identity kernels.
- use operation names for effectful dependency tracking where buffer names do
  not exist.

These are in the same problem area as #183803 and #183804.

## Tests

The updated tests cover:

- SAC CPU-offload policies automatically enabling compile-time activation
  offload.
- clear `MUST_CPU_OFFLOAD` errors for unsupported cases.
- safe `PREFER_CPU_OFFLOAD` fallback.
- eager fallback correctness.
- multi-step loss and elementwise per-parameter gradient checks.
- fake tensor behavior for AO ops.
- missing-wait and double-wait registry errors.
- multiple offloaded tensors.
- bounded forward wait placement.
- structural one-step backward reload placement.

Local lint passed:

```text
lintrunner -a
```

## Local Validation

Large nanoGPT-like runs were used for profiling and correctness on 120GB GH200:

```text
nlayer=12, dmodel=1536, nhead=12, seqlen=1024, batch=64, bf16
```

Twenty-step correctness checks passed for qkv, up_proj, and full-block targets
against baseline, including loss equality and elementwise per-parameter gradient
comparison.

Profiling summary:

```text
target      mode            steady_ms  peak_alloc  kernels  H2D overlap  D2H overlap
qkv         baseline            283.3      63.39G      589        -            -
qkv         selective_ac         298.3      56.64G      613        -            -
qkv         offload             285.6      56.64G      589      99.7%        99.6%

```

The cleanest comparison is qkv: offload gets the same allocated-memory saving as
selective AC, keeps the baseline main-compute kernel count, and is much closer
to baseline step time. The remaining gap is copy pressure rather than extra
compute kernels.

## Related Work

This PR connects several existing activation-offload threads:

- #177621 added the AO ops used here.
- #177627 makes the partitioner aware of offloaded tensors.
- #177628 removes the old stream-op path.
- #180874 adds eager SAC CPU-offload support.
- #183802 moves stream handling toward metadata plus allocator-safe stream
  recording.
- #183803 and #183804 address ordering issues needed by AO scheduling.
- #170284 tracks SAC compatibility with async activation offload.
- #158657 asks for checkpoint-level CPU/NVMe offload.
- #174960 discusses async activation offload and prefetch in FSDP.

This PR focuses on the compiled SAC-driven CPU activation offload path. It does
not add FSDP policy integration, NVMe offload, or a pinned-memory pool.

## Suggested Review Order

1. `torch/utils/checkpoint.py` and `torch/_functorch/partitioners.py` for policy
   flow and save/recompute/offload semantics.
2. `torch/_functorch/_activation_offloading/offload_ops.py` for runtime stream
   and lifetime correctness.
3. `torch/_functorch/_activation_offloading/activation_offloading.py` for wait
   sinking, reload scheduling, and graph validation.
4. Inductor ordering changes in `control_dependencies.py`, `reinplace.py`,
   `lowering.py`, and `_aot_autograd/streams.py`.
5. `test/dynamo/test_activation_offloading.py` for expected behavior and
   regression coverage.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by alint77** [comment]
qkv target:

Screenshot 2026-05-24 at 01 30 01

**Comment by alint77** [comment]
The motivation behind this feature for me is to enable longer context training, I work on protein/genome LM training optimization on JSC Jupiter cluster which has GH200 nodes. GH200/GB200/GB300 have Nvlink-C2C which provides 8x higher cpu<->gpu BW compared to PCIE, plus they have a Copy Engine which enables zero overhead H2D and D2H, fully overlap-able with compute (unlike SM copy).  One of the many ways this can help training is in activation offloading... since our models are usually tiny but we have large batch*seq dim that meant activations dominate mem usage, activation offloading is a massive win in this regime in terms of memory savings, enabling larger batch-size/seqlen which is super useful for genomeLMs

**Comment by jansel** [comment]
@claude review these changes

**Review by jansel — CHANGES_REQUESTED** [review]
fix merge conflicts

**Comment by stas00** [comment]
Thank you for working on this, @alint77.

One potential issue I see here is that this API requires users to modify the modeling code. Ideally we want this feature to work with HF models, which means unless HF transformers decides to integrate this new feature as a flag of sorts to `from_pretrained()` this feature would be very difficult to activate.

If you get a conceptual green light from the pytorch devs perhaps it might be a good idea to ping the HF transformers dev team and see if this works for them or perhaps they would propose some mods to the API?
